lib/Dialect/FIRRTL/Transforms/ProbesToSignals.cpp

```diff
@@ -0,0 +1,563 @@
+  }
+
+  // Update body, if present.
+  if (block &&
+      block
+          ->walk<mlir::WalkOrder::PreOrder>(
+              [&](Operation *op) -> WalkResult { return dispatchVisitor(op); })
+          .wasInterrupted())
+    return failure();
+
+  // Update signature and argument types.
+  if (portsToChange) {
+    mod->setAttr(mod.getPortTypesAttrName(),
+                 ArrayAttr::get(mod->getContext(), portTypes));
+
+    if (block) {
+      // We may also need to update the types on the block arguments.
+      for (auto [arg, typeAttr] :
+           llvm::zip_equal(block->getArguments(), portTypes))
+        arg.setType(cast<TypeAttr>(typeAttr).getValue());
+
+      // Drop the port stand-ins and RAUW to the block arguments.
+      for (auto [idx, wire] : wires) {
+        auto arg = block->getArgument(idx);
+        wire.getData().replaceAllUsesWith(arg);
+        wire.erase();
+      }
+    }
+  }
+
+  // Delete operations that were converted.
+  for (auto *op : llvm::reverse(toDelete))
+    op->erase();
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Visitor: Type conversion
+//===----------------------------------------------------------------------===//
+
+FailureOr<Type> ProbeVisitor::convertType(Type type, Location loc) {
+  auto err = [type, loc](const Twine &message) {
+    return mlir::emitError(loc, message) << ", cannot convert type " << type;
+  };
+  if (isa<OpenBundleType, OpenVectorType>(type))
+    return err("open aggregates not supported");
+
+  auto refType = dyn_cast<RefType>(type);
+  if (!refType)
+    return Type();
+
+  if (refType.getForceable())
+    return err("rwprobe not supported");
+
+  if (refType.getLayer())
+    return err("layer-colored probes not supported");
+
+  // Otherwise, this maps to the probed type.
+  return refType.getType();
+}
+
+//===----------------------------------------------------------------------===//
+// Visitor: Unhandled
+//===----------------------------------------------------------------------===//
+
+LogicalResult ProbeVisitor::visitUnhandledOp(Operation *op) {
+  auto checkType = [&](auto type) -> bool {
+    // Return if conversion needed (or if error).
+    auto newType = convertType(type, op->getLoc());
+    if (failed(newType))
+      return true;
+    if (!*newType)
+      return false;
+
+    // Type found that needs to be converted, diagnose.
+    op->emitError("unhandled operation needs conversion of type ")
+        << type << " to " << *newType;
+    return true;
+  };
+
+  return success(llvm::none_of(op->getOperandTypes(), checkType) &&
+                 llvm::none_of(op->getResultTypes(), checkType));
+}
+
+//===----------------------------------------------------------------------===//
+// Visitor: CHIRRTL
+//===----------------------------------------------------------------------===//
+LogicalResult
+ProbeVisitor::visitMemoryDebugPortOp(chirrtl::MemoryDebugPortOp op) {
```

Comment:
Is there any reason we can't rely on `LowerChirrtl`, or you donot want to introduce the dependency ?

Reply:
This is a bit weird, good work sniffing that out 😉 .  However, once it was implemented I didn't immediately feel compelled to rip it back out (but maybe should).

I'm not sure where we want this pass -- on one hand, supporting MemoryDebugPortOp allows it to run close to (immediately after) where LowerAnnotations is, and specifically allows it to replace the noRefTypePorts option.

That thinking is what motivated adding support for this debug port op, as it's something that option is able to handle presently.  A previous iteration showed how using this pass instead of that option did what was expected for memtaps, re-using an existing test.

However, we can't do complete conversion at that point -- due to layers and in some cases when blocks.  So the pass was moved after layer specialization.  Where these operations surely no longer exist.

But by the time those are resolved (/specialized), our pipeline has sent memories through a few passes, and without checking maybe a few more passes follow (?).

Teaching this to handle debug ports of memories (I guess those that aren't already converted to registers), is an option but offhand I'm not sure if that's a practical/plausible thing to reliably do -- but let's chat a bit!  I'm not very familiar with these operations' semantics, especially at various points in the pipeline.
  
FWIW some folks have suggested memTap's aren't needed or can/will be done differently anyway.  WDYT?  If we want to not use annotations for taps, then we'll need a story there.

I'm not worried about the dependency on LowerCHIRRTL, but good question thank you 👍 .

Reply:
Thanks, it makes sense, no need to remove this, also important to keep this as reference for however we decide to handle memtap/mem-debug-ports later.

Reply:
Sounds good! Added test back by sending it through this pass just to keep that under test in a more end-to-end way.